File: client/mia_companion/lib/utils/human_presence.dart
import 'dart:math';

/// Realistic delays for Mia's chat presence indicators (typing / recording).
class HumanPresence {
  HumanPresence._();

  static final _random = Random();

  /// How long Mia should appear to type before showing a text reply.
  /// Tuned for slower, more human-feeling mobile chat (~12–14 chars/sec).
  static Duration typingDuration(String text) {
    final chars = text.trim().length;
    if (chars == 0) return const Duration(milliseconds: 2400);

    const baseMs = 1100;
    const msPerChar = 72;
    const minMs = 2200;
    const maxMs = 32000;

    return _duration(
      baseMs: baseMs,
      msPerChar: msPerChar,
      units: chars,
      minMs: minMs,
      maxMs: maxMs,
    );
  }

  /// How long Mia should appear to record before sending a voice reply.
  /// Slower than typing — matches speaking a message aloud.
  static Duration recordingDuration(String spokenText) {
    final chars = spokenText.trim().length;
    if (chars == 0) return const Duration(milliseconds: 1800);

    const baseMs = 900;
    const msPerChar = 58;
    const minMs = 1600;
    const maxMs = 22000;

    return _duration(
      baseMs: baseMs,
      msPerChar: msPerChar,
      units: chars,
      minMs: minMs,
      maxMs: maxMs,
    );
  }

  /// Wait only the time still needed after [elapsed] (e.g. API fetch time).
  static Future<void> waitRemaining(
    Duration target,
    Stopwatch elapsed,
  ) async {
    final remaining = target - elapsed.elapsed;
    if (remaining > Duration.zero) {
      await Future.delayed(remaining);
    }
  }

  static Duration _duration({
    required int baseMs,
    required int msPerChar,
    required int units,
    required int minMs,
    required int maxMs,
  }) {
    final jitter = 0.9 + _random.nextDouble() * 0.2;
    final ms = ((baseMs + units * msPerChar) * jitter).round();
    return Duration(milliseconds: ms.clamp(minMs, maxMs));
  }
}

File: client/mia_react_native/src/utils/humanPresence.ts
function durationMs(
  baseMs: number,
  msPerChar: number,
  units: number,
  minMs: number,
  maxMs: number,
): number {
  const jitter = 0.9 + Math.random() * 0.2;
  const ms = Math.round((baseMs + units * msPerChar) * jitter);
  return Math.min(maxMs, Math.max(minMs, ms));
}

export function typingDuration(text: string): number {
  const chars = text.trim().length;
  if (chars === 0) return 1200;
  return durationMs(550, 36, chars, 1100, 16000);
}

export function recordingDuration(spokenText: string): number {
  const chars = spokenText.trim().length;
  if (chars === 0) return 1800;
  return durationMs(900, 58, chars, 1600, 22000);
}

export async function waitRemaining(targetMs: number, startedAt: number): Promise<void> {
  const remaining = targetMs - (Date.now() - startedAt);
  if (remaining > 0) {
    await new Promise((r) => setTimeout(r, remaining));
  }
}
